### ISSUE TITLE
VFS Data Connection timeout cannot be configured

### ISSUE THREAD

**Issue by nbtjlr** [wso2/product-ei#2167]
**Description:**
When using VFS transport to send file using FTP, the data connection timeout cannot be configured.
When an Data connection is not establish the client hangs and never resumes. 
This occurs rarely, but when it does the thread stays in a runnable state waiting for the data connection never gets released. 

in the fix https://wso2.org/jira/browse/ESBJAVA-3980 several changes were made regarding the definition of some of the timeout parameters. Nevertheless the data connection timeout wasn't touched. I tested a code change with success to fix this. 

**Suggested Labels:**
vfs ftp data connection timeout 

**Affected Product Version:**
EI6.2 

**Steps to reproduce:**
- Create proxy with a send mediator using VFS FTP address. 
- increase the log level on VFS FTP classes
- execute the proxy and validate the log entries. It is required to simulate an issue when the data connection is establish, in my case, we have I had a ftp server that was failing on the data connection. This can also occur if the port used for the data connection is blocked by the firewall. 

The following log entries are experienced: 
TID: [-1234] [] [2018-05-08 17:27:21,126] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  STOR QA/Sample/Connection_Test.csv
TID: [-1234] [] [2018-05-08 17:27:21,127] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  125 Data connection already open; Transfer starting.
TID: [-1234] [] [2018-05-08 17:27:21,128] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  226 Transfer complete.
TID: [-1234] [] [2018-05-08 17:27:21,128] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  PORT 10,232,160,191,163,206
TID: [-1234] [] [2018-05-08 17:27:21,129] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  200 PORT command successful.
TID: [-1234] [] [2018-05-08 17:27:21,129] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  LIST QA/Sample
TID: [-1234] [] [2018-05-08 17:27:21,129] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  125 Data connection already open; Transfer starting.
TID: [-1234] [] [2018-05-08 17:27:21,129] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  226 Transfer complete.
TID: [-1234] [] [2018-05-08 17:27:21,130] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  PORT 10,232,160,191,163,174
TID: [-1234] [] [2018-05-08 17:27:21,130] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  200 PORT command successful.
TID: [-1234] [] [2018-05-08 17:27:21,130] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  STOR QA/Sample/Connection_Test.csv
TID: [-1234] [] [2018-05-08 17:27:21,136] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  125 Data connection already open; Transfer starting.
TID: [-1234] [] [2018-05-08 17:27:21,137] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  226 Transfer complete.
TID: [-1234] [] [2018-05-08 17:27:21,137] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  PORT 10,232,160,191,146,215
TID: [-1234] [] [2018-05-08 17:27:21,138] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  200 PORT command successful.
TID: [-1234] [] [2018-05-08 17:27:21,138] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  LIST QA/Sample
TID: [-1234] [] [2018-05-08 17:27:21,138] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  150 Opening BINARY mode data connection.

If the ftp client goes into the 150 code, it will wait for the data connection to be established. If it doesnt get established it stays there forever. 
The thread on the JVM stays like: 

"vfs-Worker-19" #420 prio=5 os_prio=0 tid=0x00007f36f4022800 nid=0x3b45 runnable [0x00007f36e31b6000]
   java.lang.Thread.State: RUNNABLE
	at java.net.PlainSocketImpl.socketAccept(Native Method)
	at java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:409)
	at java.net.ServerSocket.implAccept(ServerSocket.java:545)
	at java.net.ServerSocket.accept(ServerSocket.java:513)
	at org.apache.commons.net.ftp.FTPClient._openDataConnection_(FTPClient.java:862)
	at org.apache.commons.net.ftp.FTPClient._storeFileStream(FTPClient.java:718)
	at org.apache.commons.net.ftp.FTPClient.__storeFileStream(FTPClient.java:705)
	at org.apache.commons.net.ftp.FTPClient.storeFileStream(FTPClient.java:2067)
	at org.apache.commons.vfs2.provider.ftp.FTPClientWrapper.storeFileStream(FTPClientWrapper.java:283)
	at org.apache.commons.vfs2.provider.ftp.FtpFileObject.doGetOutputStream(FtpFileObject.java:531)
	at org.apache.commons.vfs2.provider.AbstractFileObject.getOutputStream(AbstractFileObject.java:1193)
	at org.apache.commons.vfs2.provider.DefaultFileContent.getOutputStream(DefaultFileContent.java:413)
	at org.apache.synapse.transport.vfs.VFSTransportSender.populateResponseFile(VFSTransportSender.java:365)
	at org.apache.synapse.transport.vfs.VFSTransportSender.writeFile(VFSTransportSender.java:322)
	at org.apache.synapse.transport.vfs.VFSTransportSender.sendMessage(VFSTransportSender.java:197)

I tested a fix for this by changing the FTPFileProvider on the doCreateFileSystem method. I used the code created on the ESB-JAVA-3980 fix and added the DataTimeout call to it. 
This was only for testing so I set it up to be 10 times the value of the data timeout. I believe this should be defined using some parameter at the proxy level and then use it to define this in this class. 

if (this.defaultTimeout != null && (this.defaultTimeout > 0)) {
            builder.setConnectTimeout(fileSystemOptions, this.defaultTimeout);
            **builder.setDataTimeout(fileSystemOptions, 10*this.defaultTimeout);**
            builder.setSoTimeout(fileSystemOptions, this.defaultTimeout);
        }

After doing this change, the log looks like: 

TID: [-1234] [] [2018-05-13 22:37:40,307] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  200 PORT command successful.
TID: [-1234] [] [2018-05-13 22:37:40,308] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  STOR QA/Sample/Connection_Test.csv
TID: [-1234] [] [2018-05-13 22:37:40,308] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  125 Data connection already open; Transfer starting.
TID: [-1234] [] [2018-05-13 22:37:40,309] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  226 Transfer complete.
TID: [-1234] [] [2018-05-13 22:37:40,310] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  PORT 10,232,160,191,159,131
TID: [-1234] [] [2018-05-13 22:37:40,310] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  200 PORT command successful.
TID: [-1234] [] [2018-05-13 22:37:40,310] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  LIST QA/Sample
**TID: [-1234] [] [2018-05-13 22:37:40,311] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  150 Opening BINARY mode data connection.
TID: [-1234] [] [2018-05-13 22:38:30,327] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  QUIT
TID: [-1234] [] [2018-05-13 22:38:30,327] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  425 Cannot open data connection.**
TID: [-1234] [] [2018-05-13 22:38:30,329] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  220 Microsoft FTP Service
TID: [-1234] [] [2018-05-13 22:38:30,329] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  USER SampleQAUSER
TID: [-1234] [] [2018-05-13 22:38:30,329] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  331 Password required
TID: [-1234] [] [2018-05-13 22:38:30,329] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  PASS *** {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory}
TID: [-1234] [] [2018-05-13 22:38:30,330] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  230 User logged in.
TID: [-1234] [] [2018-05-13 22:38:30,330] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  TYPE I
TID: [-1234] [] [2018-05-13 22:38:30,330] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  200 Type set to I.
TID: [-1234] [] [2018-05-13 22:38:30,331] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  SYST
TID: [-1234] [] [2018-05-13 22:38:30,331] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  215 Windows_NT
TID: [-1234] [] [2018-05-13 22:38:30,331] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  PORT 10,232,160,191,155,93
TID: [-1234] [] [2018-05-13 22:38:30,332] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  200 PORT command successful.
TID: [-1234] [] [2018-05-13 22:38:30,332] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  LIST QA/Sample
TID: [-1234] [] [2018-05-13 22:38:30,333] DEBUG {org.apache.commons.vfs2.provider.ftp.FtpClientFactory$FtpConnectionFactory} -  125 Data connection already open; Transfer starting.

**Comment by nbtjlr** [comment]
@riyafa Another issue at the ftp send level. This one I have a solution that indicates where the problem needs to be fixed.

**Comment by Nirothipan** [comment]
@nbtjlr Since you have identified a possible code level fix for this, we encourage contributing it and sending a PR for it.
